BloomFilter/Program.cs

```diff
@@ -0,0 +1,346 @@
+//    save them to disk,
+// 4. Once the blockchain headers are synchronised use set a bloom filter and get 
+//    all blocks within a certain date range to check for relevant transactions.
+//
+// NOTE: This sample does not do any block verification and is NOT suitable for 
+// any kind of use on the main BitCoin network.
+//
+// Dependencies:
+// The program relies on NBitCoin (https://github.com/MetacoSA/NBitcoin) for the 
+// underlying BitCoin primitives.
+//
+// Hints:
+// The original prupose for this sample was to gain an understanding of the BitCoin
+// protocol. An invaluable tool for anyone attempting the same thing is WireShark
+// (https://www.wireshark.org/) which has a builtin BitCoin protocol decoder. To
+// use WireShark with the loopback adapter on Windows install Npcap (https://nmap.org/npcap/).
+//
+// The command line used for the local bitcoin full node:
+// "C:\Program Files\Bitcoin\daemon\bitcoind" -printtoconsole -datadir=f:\temp\bitcoind -server -testnet -debug=1 -bind=[::1]:18333
+//
+// Author(s):
+// Aaron Clauson (https://github.com/sipsorcery)
+//
+// History:
+// 14 Aug 2017  Aaron Clauson          Created.
+//
+// License: 
+// Public Domain
+// =============================================================================
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NBitcoin;
+using NBitcoin.Protocol;
+using log4net;
+
+namespace BitCoinTest
+{
+    class Program
+    {
+        static ILog logger = log4net.LogManager.GetLogger("default");
+        static Network _network = Network.TestNet;
+        static string _chainFile = "chaintest.data";
+
+        // Adjust the values below based on the BitCoin address that needs to be searched
+        // and the dates that the address had some transactions (check with https://testnet.blockexplorer.com/). 
+        static string _addressPrivateKey = "cR7X4Nd5WqA5mNwgX67th4Jo3K9vTTm28w8njLL9JT8hHPdbstL8";
+        static DateTimeOffset _startSearchTime = DateTimeOffset.Parse("14 Jul 2017");
+        static DateTimeOffset _endSearchTime = DateTimeOffset.Parse("29 Jul 2017");
+
+        // Bloom filter parameters.
+        static int _nElements = 64;
+        static double _falsePositiveRate = 0.0001;
+        static uint _nTweakIn = 50;
+
+        static void Main(string[] args)
+        {
+            Console.WriteLine("Press q at any time to quit.");
+
+            log4net.Config.XmlConfigurator.Configure();
+
+            var tokenSource = new CancellationTokenSource();
+            CancellationToken ct = tokenSource.Token;
+
+            Key key = Key.Parse(_addressPrivateKey, _network);
+            BitcoinPubKeyAddress addr = key.PubKey.GetAddress(_network);
+
+            var chain = new ConcurrentChain(_network);
+            Node node = null;
+
+            LoadChain(chain, ct).ContinueWith(t =>
```

Comment:
You can't use await directly in main so it would require an additional method or lambda to switch. I don't find ContinueWith to be "hacky" it does a good job of conveying that the intention is to wait for a task to complete before continuing.